src/main/kotlin/org/rust/ide/annotator/RustInvalidSyntaxAnnotator.kt

```diff
@@ -19,6 +24,44 @@ class RustInvalidSyntaxAnnotator : Annotator {
                 holder.createErrorAnnotation(o, "Visibility modifier is not allowed here")
             }
         }
+
+        override fun visitConstant(o: RustConstantElement) {
+            val title = if (o.static != null) "Static constant `${o.identifier.text}`" else "Constant `${o.identifier.text}`"
+            when (o.role) {
+                RustConstantRole.FREE -> {
+                    deny(o.default, "$title cannot have the `default` qualifier")
+                    require(o.expr, "$title must have a value", o)
+                }
+                RustConstantRole.TRAIT_CONSTANT -> {
+                    deny(o.vis, "$title cannot have the `pub` qualifier")
+                    deny(o.default, "$title cannot have the `default` qualifier")
+                    deny(o.static, "Static constants are not allowed in traits")
+                }
+                RustConstantRole.IMPL_CONSTANT -> {
+                    deny(o.static, "Static constants are not allowed in impl blocks")
+                    require(o.expr, "$title must have a value", o)
+                }
+                RustConstantRole.FOREIGN -> {
+                    deny(o.default, "$title cannot have the `default` qualifier")
+                    require(o.static, "Only static constants are allowed in extern blocks", o.const)
+                        ?: require(o.mut, "Non mutable static constants are not allowed in extern blocks", o.static, o.identifier)
+                    deny(o.expr, "Static constants in extern blocks cannot have values", o.eq, o.expr)
+                }
+            }
+        }
+
+        private fun require(el: PsiElement?, message: String, vararg highlightElements: PsiElement?): Annotation? =
+            if (el != null) null else holder.createErrorAnnotation(highlightElements.combinedRange ?: TextRange.EMPTY_RANGE, message)
+
+        private fun deny(el: PsiElement?, message: String, vararg highlightElements: PsiElement?): Annotation? =
+            if (el == null) null else holder.createErrorAnnotation(highlightElements.combinedRange ?: el.textRange, message)
+
+        private val Array<out PsiElement?>.combinedRange: TextRange?
+            get() = if (isEmpty())
```

Comment:
Unfortunately, `fold` with an empty initial range won't work as needed here because it will always produce ranges that start from (0, 0).